File: internal/domain/habit.go
package domain

import (
	"errors"
	"math/bits"
	"regexp"
	"strings"
	"time"
)

// Kind is how a habit is measured on a given day.
type Kind string

const (
	// KindCheck is a plain "did it / did not" habit.
	KindCheck Kind = "check"
	// KindCount counts occurrences, e.g. 8 glasses of water.
	KindCount Kind = "count"
	// KindTime counts minutes, e.g. 20 minutes of reading.
	KindTime Kind = "time"
	// KindDistance counts metres. Stored in metres rather than kilometres so a
	// day's value stays a whole number like every other kind; the UI formats it
	// as kilometres once it passes a thousand.
	KindDistance Kind = "distance"
)

// AllKinds is every kind there is, in the order the editor offers them. One
// list rather than a switch per question, so adding a kind is one edit.
var AllKinds = []Kind{KindCheck, KindCount, KindTime, KindDistance}

func (k Kind) Valid() bool {
	for _, known := range AllKinds {
		if k == known {
			return true
		}
	}
	return false
}

// KindInfo is everything the client has to know to treat a kind correctly.
type KindInfo struct {
	// Scale is how many stored units make one unit the reader writes.
	Scale int `json:"scale"`
	// Step is how much a single tap adds, in stored units.
	Step int `json:"step"`
	// Max is the largest value a target or a single day may hold.
	Max int `json:"max"`
	// Unit is the fixed unit, or "" where the user names their own.
	Unit string `json:"unit"`
}

// KindDescriptors is the per-kind table the state response carries.
//
// Sent to the client rather than mirrored there. These four numbers decide what
// a stored integer means — 5000 is five kilometres or five hundred repetitions
// depending on them — and a second copy in JavaScript is a copy that can drift.
// A drift here would not throw; it would quietly misread every entry a habit
// has.
func KindDescriptors() map[Kind]KindInfo {
	out := make(map[Kind]KindInfo, len(AllKinds))
	for _, k := range AllKinds {
		out[k] = KindInfo{Scale: k.Scale(), Step: k.Step(), Max: k.MaxTarget(), Unit: k.Unit()}
	}
	return out
}

// Scale is how many stored units go into one unit the reader writes.
//
// Values are whole numbers all the way down, so a kind that accepts a decimal
// place has to be kept finer than it is spelled: tenths of a count, tenths of
// a minute - six seconds - and metres for a distance written in kilometres.
func (k Kind) Scale() int {
	switch k {
	case KindCount, KindTime:
		return 10
	case KindDistance:
		return 1000
	}
	return 1
}

// MaxTarget is the largest daily target the kind accepts, in stored units. One
// global limit does not work across kinds: a whole day of minutes is an absurdly
// short distance, and 200 km would be a nonsensical number of glasses of water.
func (k Kind) MaxTarget() int {
	switch k {
	case KindCount:
		return 10000 // 1000 of whatever is being counted
	case KindTime:
		return 14400 // minutes in a day
	case KindDistance:
		return 200000 // 200 km
	}
	return 1
}

// Step is how much one tap adds by default, in stored units. Minutes and metres
// move in useful chunks rather than one at a time; every counting kind can be
// given a step of its own.
func (k Kind) Step() int {
	switch k {
	case KindTime:
		return 50 // 5 minutes
	case KindDistance:
		return 500 // half a kilometre
	case KindCount:
		return 10 // one of whatever is being counted
	}
	return 1
}

// Label is what the editor calls the kind. It lives here because the server
// writes it into messages the user reads, and a message naming a kind
// differently from the radio button next to it is a seam showing through.
func (k Kind) Label() string {
	switch k {
	case KindCheck:
		return "Check"
	case KindCount:
		return "Count"
	case KindTime:
		return "Time"
	case KindDistance:
		return "Distance"
	}
	return string(k)
}

// Unit is the fixed unit of a kind, or "" when the user names it themselves.
func (k Kind) Unit() string {
	switch k {
	case KindTime:
		return "min"
	case KindDistance:
		return "m"
	}
	return ""
}

// FrequencyKind is how often a habit is expected.
type FrequencyKind string

const (
	FreqDaily          FrequencyKind = "daily"
	FreqTimesPerWeek   FrequencyKind = "times_per_week"
	FreqWeekdays       FrequencyKind = "weekdays"
	FreqCustomInterval FrequencyKind = "custom_interval"
)

func (f FrequencyKind) Valid() bool {
	switch f {
	case FreqDaily, FreqTimesPerWeek, FreqWeekdays, FreqCustomInterval:
		return true
	}
	return false
}

// Weekdays is a bitmask with bit 0 = Monday through bit 6 = Sunday. A bitmask
// keeps the column a single integer, which matters once habits are queried in
// bulk for the overview.
type Weekdays uint8

func weekdayBit(d time.Weekday) Weekdays { return 1 << uint((int(d)+6)%7) }

func (w Weekdays) Has(d time.Weekday) bool { return w&weekdayBit(d) != 0 }
func (w Weekdays) Count() int              { return bits.OnesCount8(uint8(w)) }

// Frequency describes the schedule of a habit. Only the fields belonging to
// Kind carry meaning; the others are normalised to zero by Validate so that two
// habits with the same effective schedule compare equal.
type Frequency struct {
	Kind         FrequencyKind `json:"kind"`
	TimesPerWeek int           `json:"timesPerWeek"`
	Weekdays     Weekdays      `json:"weekdays"`
	IntervalDays int           `json:"intervalDays"`
	// WeekInterval narrows chosen weekdays to every n-th week, counted in whole
	// Monday-to-Sunday weeks from the week of AnchorDate. One is every week.
	WeekInterval int `json:"weekInterval"`
	// WeekOfMonth narrows chosen weekdays to their n-th occurrence in the month:
	// 1 to 4, or LastWeekOfMonth for the last one. Zero is every occurrence. It
	// cannot be combined with a WeekInterval above one.
	WeekOfMonth int `json:"weekOfMonth"`
	// AnchorDate is the first due day of a custom-interval schedule, or the
	// start of an every-n-weeks one. Without it the phase would silently shift
	// whenever the habit is edited.
	AnchorDate Date `json:"anchorDate"`
}

// LastWeekOfMonth is the WeekOfMonth for the last occurrence of a weekday in
// its month, whether that is the fourth or the fifth.
const LastWeekOfMonth = -1

// Habit is a tracked habit belonging to exactly one user.
type Habit struct {
	ID    string `json:"id"`
	Name  string `json:"name"`
	Color string `json:"color"`
	// Icon names one of HabitIcons, or is empty for a habit drawn without one.
	Icon string `json:"icon"`
	Kind Kind   `json:"kind"`
	// CategoryID is empty when the habit belongs to no category. It may also
	// point at a soft-deleted category, in which case the habit shows up as
	// uncategorised until that category is restored.
	CategoryID  string `json:"categoryId"`
	TargetValue int    `json:"targetValue"`
	// StepValue is how much one tap, or one press of + or -, adds. It starts at
	// the kind's natural step and every kind but a tick lets the user change
	// it: "one glass" is the obvious increment for water, "ten" is for push-ups.
	StepValue  int        `json:"stepValue"`
	Unit       string     `json:"unit"`
	Frequency  Frequency  `json:"frequency"`
	Position   int        `json:"position"`
	ArchivedAt *time.Time `json:"archivedAt"`
	CreatedAt  time.Time  `json:"createdAt"`
	UpdatedAt  time.Time  `json:"updatedAt"`
}

var (
	ErrValidation = errors.New("validation error")
	colorPattern  = regexp.MustCompile(`^#[0-9a-fA-F]{6}$`)
)

const (
	MaxNameLen = 80
	MaxUnitLen = 16
)

// The bounds messages are phrased per kind, so the reader is told about
// minutes or metres rather than about an abstract "target".
func targetTooSmall(k Kind) error {
	switch k {
	case KindTime:
		return invalid("time must be at least 0.1 minutes")
	case KindDistance:
		return invalid("distance must be at least 1 metre")
	}
	return invalid("target must be at least 0.1")
}

// tooLarge names a kind's ceiling the way the reader wrote it: the stored
// number divided back down, with the unit it is entered in. what is the thing
// being bounded - "target", "step", "value" - and is part of the template, so
// every combination is a sentence of its own for the translation to key on.
func tooLarge(what string, k Kind) error {
	limit := k.MaxTarget() / k.Scale()
	switch k {
	case KindTime:
		return invalid(what+" may be at most {max} minutes", "max", limit)
	case KindDistance:
		return invalid(what+" may be at most {max} kilometres", "max", limit)
	}
	return invalid(what+" may be at most {max}", "max", limit)
}

func targetTooLarge(k Kind) error {
	switch k {
	case KindTime:
		return tooLarge("time", k)
	case KindDistance:
		return tooLarge("distance", k)
	}
	return tooLarge("target", k)
}

var invalid = Invalid

// Validate normalises the habit in place and reports why it is unacceptable.
// Normalising here rather than in the HTTP layer means the invariants hold for
// every future entry point, including imports and CLI tooling.
func (h *Habit) Validate() error {
	h.Name = strings.TrimSpace(h.Name)
	h.Unit = strings.TrimSpace(h.Unit)

	if h.Name == "" {
		return invalid("name must not be empty")
	}
	if len([]rune(h.Name)) > MaxNameLen {
		return invalid("name is longer than {max} characters", "max", MaxNameLen)
	}
	if len([]rune(h.Unit)) > MaxUnitLen {
		return invalid("unit is longer than {max} characters", "max", MaxUnitLen)
	}
	if h.Color == "" {
		h.Color = DefaultColors[0]
	}
	if !colorPattern.MatchString(h.Color) {
		return invalid("colour must be a hex value like #4caf50")
	}
	h.Color = strings.ToLower(h.Color)

	h.Icon = strings.TrimSpace(h.Icon)
	if h.Icon != "" && !ValidIcon(h.Icon) {
		return invalid(`unknown icon "{icon}"`, "icon", h.Icon)
	}

	if !h.Kind.Valid() {
		return invalid(`unknown habit kind "{kind}"`, "kind", h.Kind)
	}
	if h.Kind == KindCheck {
		// A tick is done or it is not; there is nothing to configure.
		h.TargetValue = 1
	} else if h.TargetValue < 1 {
		return targetTooSmall(h.Kind)
	}
	if h.TargetValue > h.Kind.MaxTarget() {
		return targetTooLarge(h.Kind)
	}
	// A tick has nothing to count, so its step stays one whatever a client
	// sends. The counting kinds start at the step of their unit and may be set
	// to anything up to their own maximum.
	if h.Kind == KindCheck {
		h.StepValue = 1
	} else if h.StepValue < 1 {
		h.StepValue = h.Kind.Step()
	} else if h.StepValue > h.Kind.MaxTarget() {
		return tooLarge("step", h.Kind)
	}
	// Kinds with a fixed unit own it; only a count lets the user name one.
	if u := h.Kind.Unit(); u != "" || h.Kind == KindCheck {
		h.Unit = u
	}

	return h.normaliseFrequency()
}

func (h *Habit) normaliseFrequency() error {
	f := &h.Frequency
	if !f.Kind.Valid() {
		return invalid(`unknown frequency "{frequency}"`, "frequency", f.Kind)
	}
	switch f.Kind {
	case FreqDaily:
		f.TimesPerWeek, f.Weekdays, f.IntervalDays, f.AnchorDate = 0, 0, 0, Date{}
		f.WeekInterval, f.WeekOfMonth = 0, 0
	case FreqTimesPerWeek:
		if f.TimesPerWeek < 1 || f.TimesPerWeek > 7 {
			return invalid("times per week must be between 1 and 7")
		}
		f.Weekdays, f.IntervalDays, f.AnchorDate = 0, 0, Date{}
		f.WeekInterval, f.WeekOfMonth = 0, 0
	case FreqWeekdays:
		if f.Weekdays == 0 {
			return invalid("at least one weekday must be selected")
		}
		if f.Weekdays > 0b1111111 {
			return invalid("invalid weekday selection")
		}
		// Zero is what a client that knows nothing of the week interval sends,
		// and it means what it always has: every week.
		if f.WeekInterval == 0 {
			f.WeekInterval = 1
		}
		if f.WeekInterval < 1 || f.WeekInterval > 52 {
			return invalid("week interval must be between 1 and 52 weeks")
		}
		if f.WeekOfMonth != LastWeekOfMonth && (f.WeekOfMonth < 0 || f.WeekOfMonth > 4) {
			return invalid("week of the month must be 1 to 4 or the last")
		}
		if f.WeekInterval > 1 && f.WeekOfMonth != 0 {
			return invalid("a week interval and a week of the month cannot be combined")
		}
		// Only an every-n-weeks schedule has a phase to keep.
		if f.WeekInterval > 1 {
			if f.AnchorDate.IsZero() {
				f.AnchorDate = DateFromTime(h.CreatedAt)
			}
		} else {
			f.AnchorDate = Date{}
		}
		f.TimesPerWeek, f.IntervalDays = 0, 0
	case FreqCustomInterval:
		if f.IntervalDays < 1 || f.IntervalDays > 365 {
			return invalid("interval must be between 1 and 365 days")
		}
		if f.AnchorDate.IsZero() {
			f.AnchorDate = DateFromTime(h.CreatedAt)
		}
		f.TimesPerWeek, f.Weekdays = 0, 0
		f.WeekInterval, f.WeekOfMonth = 0, 0
	}
	return nil
}

// ValidateEntryValue bounds a single day's value.
//
// A day is held to the same per-kind ceiling as a target: the client already
// stops a tap there, but the API is reachable without it, and an unbounded
// value would overflow the sum in Stats.Total long before it meant anything.
func ValidateEntryValue(k Kind, value int) error {
	if !k.Valid() {
		return invalid(`unknown habit kind "{kind}"`, "kind", k)
	}
	if value < 0 {
		return invalid("value must not be negative")
	}
	if value > k.MaxTarget() {
		return tooLarge("value", k)
	}
	return nil
}

// Target is the value that counts as done for a single day.
func (h Habit) Target() int {
	if h.Kind == KindCheck {
		return 1
	}
	if h.TargetValue < 1 {
		return 1
	}
	return h.TargetValue
}

// IsComplete reports whether a day's value reaches the habit's target.
func (h Habit) IsComplete(value int) bool { return value >= h.Target() }

// IsScheduled reports whether the habit is due on d.
//
// FreqTimesPerWeek has no fixed days by design: any day of the week is a valid
// opportunity, and whether the week was met is a question for the weekly stats
// rather than for a single day.
func (h Habit) IsScheduled(d Date) bool {
	switch h.Frequency.Kind {
	case FreqDaily, FreqTimesPerWeek:
		return true
	case FreqWeekdays:
		return h.Frequency.Weekdays.Has(d.Weekday()) && h.inScheduledWeek(d)
	case FreqCustomInterval:
		n := h.Frequency.IntervalDays
		if n < 1 {
			return false
		}
		anchor := h.Frequency.AnchorDate
		if anchor.IsZero() {
			anchor = DateFromTime(h.CreatedAt)
		}
		diff := d.DaysSince(anchor)
		if diff < 0 {
			return false
		}
		return diff%n == 0
	}
	return false
}

// inScheduledWeek reports whether d, already on one of the chosen weekdays,
// also falls in a week the schedule asks for: every n-th week from the anchor,
// or the n-th occurrence of that weekday in its month.
func (h Habit) inScheduledWeek(d Date) bool {
	f := h.Frequency
	if f.WeekOfMonth == LastWeekOfMonth {
		return d.AddDays(7).Month != d.Month
	}
	if f.WeekOfMonth > 0 {
		return (d.Day-1)/7+1 == f.WeekOfMonth
	}
	if f.WeekInterval > 1 {
		anchor := f.AnchorDate
		if anchor.IsZero() {
			anchor = DateFromTime(h.CreatedAt)
		}
		if d.Before(anchor) {
			return false
		}
		weeks := d.StartOfWeek().DaysSince(anchor.StartOfWeek()) / 7
		return weeks%f.WeekInterval == 0
	}
	return true
}

// AcceptsEntry reports whether a value may be recorded on d.
//
// Frequencies with fixed days — chosen weekdays and a custom interval — close all
// other days: the days are the whole point of those frequencies, so a tick in
// between is a mistake. Daily and times-per-week stay open on every day, since
// any day counts for them anyway.
func (h Habit) AcceptsEntry(d Date) bool {
	switch h.Frequency.Kind {
	case FreqWeekdays, FreqCustomInterval:
		return h.IsScheduled(d)
	}
	return true
}

func (h Habit) IsArchived() bool { return h.ArchivedAt != nil }

// DefaultColors is the palette offered in the editor, kept in the domain so the
// server can validate against the same list the client renders.
var DefaultColors = []string{
	"#dc2626", "#ea580c", "#eab308", "#65a30d",
	"#16a34a", "#0d9488", "#0284c7", "#2563eb",
	"#4f46e5", "#7c3aed", "#db2777", "#64748b",
}

// HabitIcons are the icons a habit may wear, in the order the editor offers
// them. Only the names live here; the drawings are the client's. Kept on the
// server for the same reason as the palette: a name outside this list would be
// stored and then drawn as nothing at all.
var HabitIcons = []string{
	"droplet", "apple", "utensils", "coffee", "pill", "heart", "dumbbell", "bike",
	"mountain", "flame", "bed", "moon", "sun", "book", "pencil", "lightbulb",
	"code", "globe", "music", "palette", "camera", "leaf", "home", "wallet",
	"users", "smartphone", "ban", "smile", "star", "target", "clock", "check",
}

// ValidIcon reports whether name is one of HabitIcons.
func ValidIcon(name string) bool {
	for _, known := range HabitIcons {
		if name == known {
			return true
		}
	}
	return false
}

File: web/assets/js/habit.js
// Presentation helpers shared by the overview and the detail view.

import { weekdayIndex, daysBetween, addDays, startOfWeek, WEEKDAY_SHORT } from "./dates.js";
import { state } from "./state.js";

/**
 * Whether the habit is due on a given day.
 *
 * This mirrors domain.Habit.IsScheduled on the server. The duplication is
 * deliberate: the grid has to shade non-due days for a whole screen of dates,
 * and asking the server for that would turn every scroll into a round trip. The
 * server stays the authority — it is what validates writes.
 */
export function isScheduled(habit, iso) {
  const f = habit.frequency;
  switch (f.kind) {
    case "daily":
    case "times_per_week":
      return true;
    case "weekdays":
      return (f.weekdays & (1 << weekdayIndex(iso))) !== 0;
    case "every_n_days": {
      if (!f.intervalDays || !f.anchorDate) return false;
      const diff = daysBetween(f.anchorDate, iso);
      return diff >= 0 && diff % f.intervalDays === 0;
    }
    default:
      return false;
  }
}

/**
 * Whether a value may be recorded on a day. Mirrors domain.Habit.AcceptsEntry:
 * a habit with fixed days (chosen weekdays, every n days) closes the others.
 */
export function acceptsEntry(habit, iso) {
  const kind = habit.frequency.kind;
  if (kind === "weekdays" || kind === "every_n_days") return isScheduled(habit, iso);
  return true;
}

/**
 * What the server says about a kind: how fine its stored unit is, how much one
 * tap adds, and the largest a day may be.
 *
 * These numbers decide what a stored integer means — 5000 is five kilometres or
 * five hundred repetitions depending on them — so they come down with the state
 * rather than being written out a second time here. The fallbacks below only
 * cover the moment before the first response has landed; nothing draws then.
 */
const FALLBACK = {
  check: { scale: 1, step: 1, max: 1, unit: "" },
  count: { scale: 10, step: 10, max: 10000, unit: "" },
  time: { scale: 10, step: 50, max: 14400, unit: "min" },
  distance: { scale: 1000, step: 500, max: 200000, unit: "m" },
};

export function kindInfo(kind) {
  return state.kinds?.[kind] ?? FALLBACK[kind] ?? FALLBACK.check;
}

/**
 * How many stored units make one unit the reader writes.
 *
 * Every value the app stores is a whole number, so the kinds that take a
 * decimal place are kept finer than they are written: tenths of a count,
 * tenths of a minute, and metres for a distance spelled in kilometres.
 */
export function scale(habit) {
  return scaleOf(habit.kind);
}

/** The same, for a kind that has no habit yet — the editor's target fields. */
export function scaleOf(kind) {
  return kindInfo(kind).scale;
}

/** A stored number written out, with at most one decimal place. */
function written(habit, value) {
  return (value / scale(habit)).toLocaleString("en-GB", { maximumFractionDigits: 1 });
}

export function target(habit) {
  return habit.kind === "check" ? 1 : Math.max(1, habit.targetValue || 1);
}

export function isComplete(habit, value) {
  return (value || 0) >= target(habit);
}

/** 0…1, for the progress ring. */
export function progress(habit, value) {
  return Math.max(0, Math.min(1, (value || 0) / target(habit)));
}

/**
 * How much a single tap adds.
 *
 * A count carries its own step, because "one" is right for glasses of water
 * and wrong for push-ups. The other kinds follow the unit they are stored in:
 * minutes move in fives and metres in half-kilometres — tapping a 5 km run up
 * in single metres would be absurd. The long-press dialog is there for an
 * exact value.
 */
export function step(habit) {
  if (habit.stepValue > 0) return habit.stepValue;
  return kindInfo(habit.kind).step;
}

/**
 * The value a tap writes.
 *
 * Counting kinds keep going past their target: after 5 km, one more tap means
 * 5.5 km, not "start over". Overshooting the target is ordinary — the day the
 * run was longer — while wanting to wipe a day is rare, and the long press and
 * the right click set any value including zero.
 *
 * The only ceiling is the per-kind maximum, the same one the value dialog uses,
 * so a held-down tap cannot write a number the habit could never have as its
 * target.
 */
export function nextValue(habit, current) {
  if (habit.kind === "check") return current > 0 ? 0 : 1;
  return Math.min((current || 0) + step(habit), maxValue(habit));
}

export function unitLabel(habit) {
  return kindInfo(habit.kind).unit || habit.unit || "";
}

/** Metres as the reader would say them: 800 m, 5 km, 12.5 km. */
export function formatDistance(metres) {
  if (metres < 1000) return `${metres} m`;
  return `${(Math.round(metres / 100) / 10).toLocaleString("en-GB")} km`;
}

/**
 * The short form that fits inside a day circle — no unit, since the circle is
 * about 25px across. Distances are shown in kilometres, because five digits of
 * metres would never fit.
 */
export function cellValue(habit, value) {
  // A kilometre with one decimal, whatever the metres say: 5.2 fits, 5200 and
  // 5.24 do not.
  if (habit.kind === "distance") return (Math.round(value / 100) / 10).toLocaleString("en-GB");
  return written(habit, value);
}

/** A single day's value spelled out with its unit, for tooltips and labels. */
export function formatValue(habit, value) {
  switch (habit.kind) {
    case "distance": return formatDistance(value);
    case "time": return `${written(habit, value)} min`;
    case "count": {
      const n = written(habit, value);
      return habit.unit ? `${n} ${habit.unit}` : `${n}×`;
    }
    default: return `${value}×`;
  }
}

/**
 * A sum spelled out, which is not always the same as a single day's value:
 * 225 minutes read over a year reads better as "3 h 45 min".
 */
export function formatTotal(habit, total) {
  switch (habit.kind) {
    case "time": {
      const minutes = total / scale(habit);
      const hours = Math.floor(minutes / 60);
      if (hours === 0) return `${formatMinutes(minutes)} min`;
      return `${hours} h ${formatMinutes(minutes % 60)} min`;
    }
    case "distance":
      return formatDistance(total);
    case "count": {
      const n = written(habit, total);
      return habit.unit ? `${n} ${habit.unit}` : `${n}×`;
    }
    default:
      return `${total}×`;
  }
}

const formatMinutes = (minutes) =>
  minutes.toLocaleString("en-GB", { maximumFractionDigits: 1 });

/** Habits whose values are a quantity worth adding up. */
export function isCountable(habit) {
  return habit.kind === "count" || habit.kind === "time" || habit.kind === "distance";
}

/**
 * A stretch of history summed into buckets of one day, one week or one month.
 *
 * Computed from the entries the client already holds rather than asked of the
 * server — the detail view has pulled the habit's full history anyway, and a
 * sum over a few hundred numbers is not worth a round trip.
 *
 * Every bucket carries both numbers the chart draws: `sum` is what that day or
 * week or month brought in, `cumulative` the running total since the start of
 * the range.
 */
export function periodSummary(habit, from, to, size) {
  // Weeks are counted from the Monday of the week the range starts in, so a
  // bucket is never half a week wide.
  const origin = size === "week" ? startOfWeek(from) : from;
  const indexOf = (iso) => {
    if (size === "day") return daysBetween(origin, iso);
    if (size === "week") return Math.floor(daysBetween(origin, iso) / 7);
    return monthsBetween(origin, iso);
  };

  const buckets = [];
  for (let i = 0; i <= indexOf(to); i++) {
    buckets.push({ start: startOfBucket(origin, i, size), sum: 0, cumulative: 0 });
  }

  let total = 0;
  let best = 0;
  let activeDays = 0;
  for (const [iso, value] of Object.entries(habit.entries)) {
    if (iso < from || iso > to || value <= 0) continue;
    buckets[indexOf(iso)].sum += value;
    total += value;
    activeDays++;
    if (value > best) best = value;
  }

  let running = 0;
  for (const bucket of buckets) {
    running += bucket.sum;
    bucket.cumulative = running;
  }
  return { buckets, total, best, activeDays };
}

/** The first day of the i-th bucket after `origin`. */
function startOfBucket(origin, i, size) {
  if (size === "day") return addDays(origin, i);
  if (size === "week") return addDays(origin, i * 7);
  const month = Number(origin.slice(5, 7)) - 1 + i;
  const year = Number(origin.slice(0, 4)) + Math.floor(month / 12);
  return `${year}-${String((month % 12) + 1).padStart(2, "0")}-01`;
}

function monthsBetween(from, to) {
  return (Number(to.slice(0, 4)) - Number(from.slice(0, 4))) * 12 +
    Number(to.slice(5, 7)) - Number(from.slice(5, 7));
}

export function describeTarget(habit) {
  switch (habit.kind) {
    case "count":
      return `${written(habit, habit.targetValue)}${habit.unit ? " " + habit.unit : "×"}`;
    case "time":
      return `${written(habit, habit.targetValue)} min`;
    case "distance":
      return formatDistance(habit.targetValue);
    default:
      return "";
  }
}

export function describeFrequency(habit) {
  const f = habit.frequency;
  switch (f.kind) {
    case "daily":
      return "daily";
    case "times_per_week":
      return `${f.timesPerWeek}× per week`;
    case "weekdays": {
      // Abbreviated: five spelled-out names run past the name column, and the
      // short forms are unambiguous in a line that is about a weekly rhythm.
      const days = WEEKDAY_SHORT.filter((_, i) => f.weekdays & (1 << i));
      if (days.length === 7) return "daily";
      if (days.length === 5 && !(f.weekdays & (1 << 5)) && !(f.weekdays & (1 << 6))) {
        return "Mon–Fri";
      }
      return days.join(", ");
    }
    case "every_n_days":
      return f.intervalDays === 1 ? "daily" : `every ${f.intervalDays} days`;
    default:
      return "";
  }
}

/** The subtitle under a habit name: streak first, then what is expected. */
export function describeHabit(habit) {
  const parts = [];
  // Said first, because a dimmed row alone is easy to misread as merely faded.
  if (habit.archivedAt) parts.push("Archived");
  // No streak here for now: with the target and the frequency it made the line
  // longer than the name column, and it is still readable on the detail view
  // under "Current streak". The stat itself is untouched and waiting for a
  // better place on the board.
  // Frequency first, target second: how often is the habit's shape, the target
  // the detail of a single day — and the frequency is the shorter of the two,
  // so the part that survives a narrow column is the one that says more.
  const t = describeTarget(habit);
  parts.push(t ? `${describeFrequency(habit)} · ${t}` : describeFrequency(habit));
  return parts.join(" · ");
}

/**
 * The streak levels the board paints, as the number of calendar days a run has
 * to have been going before a day earns that level.
 *
 * Stated in days rather than in the habit's own rhythm, because the steps are
 * milestones a person recognises — a week, a fortnight, a month, a quarter,
 * half a year, a year — and those pass at the same rate whether a habit is due
 * daily or three times a week.
 *
 * Here rather than on the server for the same reason as heatLevel above: the
 * server decides what a run *is*, this decides only how it is drawn.
 */
export const STREAK_LEVELS = [7, 14, 30, 90, 180, 365];

/** Level 0…6 for a run that has been going for `days` calendar days. */
export function streakLevel(days) {
  let level = 0;
  for (const needed of STREAK_LEVELS) {
    if (days >= needed) level++;
  }
  return level;
}

/**
 * How many calendar days the run covering `iso` had been going by that day, or
 * 0 for a day outside every run.
 *
 * Counted from the run's own first day, which the server sends even when it is
 * older than the history the board holds: a year-long streak is painted as one
 * the moment it is a year old, not once its start scrolls into view.
 */
export function streakDaysOn(habit, iso) {
  for (const run of habit.streakRuns ?? []) {
    // ISO dates sort as text, which is what makes this a comparison and not a
    // date calculation per cell.
    if (iso < run.from || iso > run.to) continue;
    return daysBetween(run.from, iso) + 1;
  }
  return 0;
}

/** Heat level 0…4 for the calendar heatmap. */
export function heatLevel(habit, value) {
  if (!value) return 0;
  const p = progress(habit, value);
  if (p >= 1) return 4;
  if (p >= 0.66) return 3;
  if (p >= 0.33) return 2;
  return 1;
}

/** The largest value a single day may hold. The server enforces the same cap. */
export function maxValue(habit) {
  return kindInfo(habit.kind).max;
}
